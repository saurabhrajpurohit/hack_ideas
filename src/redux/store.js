import { createStore } from "redux";
import { SET_USER } from "../utils/constants";

const userState = {};

const userReducer = (state = userState, action) => {
    switch (action.type) {
        case SET_USER:
            return {
                ...state,
                user: action.value
            }
        default:
            return state;
    }
};

export default createStore(userReducer);

