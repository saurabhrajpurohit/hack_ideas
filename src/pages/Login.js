import React from "react";
import { Box, Stack, Typography, Button, Container, Link, TextField, Alert, LinearProgress } from "@mui/material";
import { styled } from '@mui/material/styles';
import { UserService } from "../service/Service";
import AuthLayout from "../layouts/AuthLayout";
import { Link as RouterLink, useNavigate } from 'react-router-dom';
import { useDispatch } from "react-redux";
import { setUser } from "../redux/actions";

const ContentStyle = styled('div')(({ theme }) => ({
    maxWidth: 480,
    margin: 'auto',
    display: 'flex',
    minHeight: '100vh',
    flexDirection: 'column',
    justifyContent: 'center',
    padding: theme.spacing(12, 0)
}));

const Login = () => {
    const navigate = useNavigate();
    const dispatch = useDispatch();
    const [employeeID, setEmployeeID] = React.useState("");
    const [fieldError, setFieldError] = React.useState();
    const [loading, setLoading] = React.useState(false);
    const [error, setError] = React.useState(null);

    function isValid() {
        if (!employeeID) {
            setFieldError("Employee ID is required");
            return false;
        } else if (isNaN(employeeID)) {
            setFieldError("Employee ID must be a number");
            return false;
        }
        return true;
    };

    function submit(event) {
        event.preventDefault();
        if (isValid()) {
            setLoading(true);
            UserService.getByField("EmployeeID", Number(employeeID)).then(data => {
                setLoading(false);
                console.log(data);
                if (data.length) {
                    dispatch(setUser(data[0]));
                    navigate("/hack/list");
                } else {
                    setError("Employee ID is not registered in our system.");
                }
            }).catch(err => {
                setLoading(false);
                console.error(err);
                setError("We are unable to login at this time. Please try again later.");
            });
        }
    };

    return (
        <Box data-testid="login-form">
            <AuthLayout>
                Donâ€™t have an account? &nbsp;
                <Link
                    underline="none"
                    variant="subtitle2"
                    component={RouterLink}
                    to="/signup"
                >
                    Get started
                </Link>
            </AuthLayout>
            <Container maxWidth="sm">
                <ContentStyle>
                    <form onSubmit={submit}>
                        <Stack sx={{ mb: 5 }}>
                            <Typography variant="h4" gutterBottom>
                                Sign in to Hack- Ideas
                            </Typography>
                            <Typography sx={{ color: 'text.secondary' }}>
                                Enter your Employee ID below.
                            </Typography>
                        </Stack>
                        {error && <Box sx={{ width: '100%' }} mb={1}>
                            <Alert severity="error">{error}</Alert>
                        </Box>}
                        <Box mb={2}>
                            <Alert severity="info">
                                Use Employee ID: <strong>1001, 1002, 1003, 1004</strong> to Enter.
                            </Alert>
                        </Box>
                        <Stack spacing={3}>
                            <TextField
                                id="employeeID"
                                value={employeeID}
                                variant="outlined"
                                label="Employee ID"
                                error={!!fieldError}
                                helperText={fieldError}
                                inputProps={{
                                    "data-testid":"employee-id"
                                }}
                                onChange={(event) => setEmployeeID(event.target.value)}
                            />
                            {loading && <Box sx={{ width: '100%' }}>
                                <LinearProgress />
                            </Box>}
                            <Button
                                id="submit"
                                variant="contained"
                                color="primary"
                                type="submit"
                                data-testid="login-submit"
                            >
                                Login
                            </Button>
                        </Stack>
                    </form>
                </ContentStyle>
            </Container>
        </Box>
    )
}

export default Login;