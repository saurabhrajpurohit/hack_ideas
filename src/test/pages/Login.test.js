import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import Login from '../../pages/Login';
import store from '../../redux/store';

test('ui to be rendered correctly', () => {
    render(
        <Provider store={store}>
            <BrowserRouter>
                <Login />
            </BrowserRouter>
        </Provider>
    );
    const element = screen.getByTestId('login-form');
    expect(element).toMatchSnapshot();
});

test('ui to be rendered with errors', () => {
    render(
        <Provider store={store}>
            <BrowserRouter>
                <Login />
            </BrowserRouter>
        </Provider>
    );
    const button = screen.getByTestId('login-submit');
    fireEvent.click(button);
    const element = screen.getByTestId('login-form');
    expect(element).toMatchSnapshot();
});