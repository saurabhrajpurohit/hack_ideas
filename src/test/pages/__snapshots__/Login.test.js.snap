// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui to be rendered correctly 1`] = `
<div
  class="MuiBox-root css-0"
  data-testid="login-form"
>
  <header
    class="css-sg2wnk"
  >
    <a
      href="/"
    >
      <div
        class="MuiBox-root css-1yb5pbx"
      >
        <span
          class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom css-1hleqvg-MuiTypography-root"
        >
          Hack 
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EmojiObjectsIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 3c-.46 0-.93.04-1.4.14-2.76.53-4.96 2.76-5.48 5.52-.48 2.61.48 5.01 2.22 6.56.43.38.66.91.66 1.47V19c0 1.1.9 2 2 2h.28c.35.6.98 1 1.72 1s1.38-.4 1.72-1H14c1.1 0 2-.9 2-2v-2.31c0-.55.22-1.09.64-1.46C18.09 13.95 19 12.08 19 10c0-3.87-3.13-7-7-7zm2 16h-4v-1h4v1zm0-2h-4v-1h4v1zm-1.5-5.59V14h-1v-2.59L9.67 9.59l.71-.71L12 10.5l1.62-1.62.71.71-1.83 1.82z"
            />
          </svg>
           Idea's
        </span>
      </div>
    </a>
    <p
      class="MuiTypography-root MuiTypography-body2 css-p34ji4-MuiTypography-root"
    >
      Don’t have an account?  
      <a
        class="MuiTypography-root MuiTypography-subtitle2 MuiLink-root MuiLink-underlineNone css-1ser5cc-MuiTypography-root-MuiLink-root"
        href="/signup"
      >
        Get started
      </a>
    </p>
  </header>
  <div
    class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
  >
    <div
      class="css-26q4qm"
    >
      <form>
        <div
          class="css-lr1r4n-MuiStack-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
          >
            Sign in to Hack- Ideas
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
          >
            Enter your Employee ID below.
          </p>
        </div>
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-12mvehg-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="InfoOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-acap47-MuiAlert-message"
            >
              Use Employee ID: 
              <strong>
                1001, 1002, 1003, 1004
              </strong>
               to Enter.
            </div>
          </div>
        </div>
        <div
          class="css-1lwbda4-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="employeeID"
              id="employeeID-label"
            >
              Employee ID
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="employee-id"
                id="employeeID"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Employee ID
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
            data-testid="login-submit"
            id="submit"
            tabindex="0"
            type="submit"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`ui to be rendered with errors 1`] = `
<div
  class="MuiBox-root css-0"
  data-testid="login-form"
>
  <header
    class="css-sg2wnk"
  >
    <a
      href="/"
    >
      <div
        class="MuiBox-root css-1yb5pbx"
      >
        <span
          class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom css-1hleqvg-MuiTypography-root"
        >
          Hack 
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EmojiObjectsIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 3c-.46 0-.93.04-1.4.14-2.76.53-4.96 2.76-5.48 5.52-.48 2.61.48 5.01 2.22 6.56.43.38.66.91.66 1.47V19c0 1.1.9 2 2 2h.28c.35.6.98 1 1.72 1s1.38-.4 1.72-1H14c1.1 0 2-.9 2-2v-2.31c0-.55.22-1.09.64-1.46C18.09 13.95 19 12.08 19 10c0-3.87-3.13-7-7-7zm2 16h-4v-1h4v1zm0-2h-4v-1h4v1zm-1.5-5.59V14h-1v-2.59L9.67 9.59l.71-.71L12 10.5l1.62-1.62.71.71-1.83 1.82z"
            />
          </svg>
           Idea's
        </span>
      </div>
    </a>
    <p
      class="MuiTypography-root MuiTypography-body2 css-p34ji4-MuiTypography-root"
    >
      Don’t have an account?  
      <a
        class="MuiTypography-root MuiTypography-subtitle2 MuiLink-root MuiLink-underlineNone css-1ser5cc-MuiTypography-root-MuiLink-root"
        href="/signup"
      >
        Get started
      </a>
    </p>
  </header>
  <div
    class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
  >
    <div
      class="css-26q4qm"
    >
      <form>
        <div
          class="css-lr1r4n-MuiStack-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1vw6mcs-MuiTypography-root"
          >
            Sign in to Hack- Ideas
          </h4>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
          >
            Enter your Employee ID below.
          </p>
        </div>
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-12mvehg-MuiPaper-root-MuiAlert-root"
            role="alert"
          >
            <div
              class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="InfoOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message css-acap47-MuiAlert-message"
            >
              Use Employee ID: 
              <strong>
                1001, 1002, 1003, 1004
              </strong>
               to Enter.
            </div>
          </div>
        </div>
        <div
          class="css-1lwbda4-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="employeeID"
              id="employeeID-label"
            >
              Employee ID
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-describedby="employeeID-helper-text"
                aria-invalid="true"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="employee-id"
                id="employeeID"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Employee ID
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
              id="employeeID-helper-text"
            >
              Employee ID is required
            </p>
          </div>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
            data-testid="login-submit"
            id="submit"
            tabindex="0"
            type="submit"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
